integration-tests/tests/network/runner.rs

```diff
@@ -320,25 +356,35 @@ impl StateMachine {
                                 .map_err(|_| ())
                                 .and_then(move |res| {
                                     if let NetworkResponses::PingPongInfo { pings, pongs } = res {
-                                        let len_matches = pings.len() == pings_expected.len()
-                                            && pongs.len() == pongs_expected.len();
-
-                                        let ping_ok = len_matches
-                                            && pings_expected.into_iter().all(|(nonce, source)| {
-                                                pings
-                                                    .get(&nonce)
-                                                    .map_or(false, |ping| ping.source == source)
-                                            });
-
-                                        let pong_ok = len_matches
-                                            && pongs_expected.into_iter().all(|(nonce, source)| {
-                                                pongs
-                                                    .get(&nonce)
-                                                    .map_or(false, |pong| pong.source == source)
-                                            });
-
-                                        if len_matches && ping_ok && pong_ok {
+                                        let ping_ok = pings.len() == pings_expected.len()
+                                            && pings_expected.into_iter().all(
+                                                |(nonce, source, count)| {
+                                                    pings.get(&nonce).map_or(false, |ping| {
+                                                        ping.0.source == source
+                                                            && (count.is_none()
+                                                                || count.unwrap() == ping.1)
+                                                    })
+                                                },
+                                            );
```

Comment:
That's a bussy snippet of code, though I don't see a clear way to write this better. Probably OK for tests.